flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java

```diff
@@ -216,7 +216,7 @@
     protected final Properties producerConfig;
 
     /** The name of the default topic this producer is writing data to. */
-    protected final String defaultTopicId;
+    @Nullable protected String defaultTopicId;
```

Comment:
I don't understand why we need to make default topic nullable. We could just set "unspecified" on table API if there is no other value. Potentially add some special character that kafka doesn't support to make sure writing fails.
As is, we change quite a bit of API and semantics and introduce new bugs (there is no guard against `defaultTopicId = null` and `getTargetTopic = null`).
Another idea would be to explicitly support default topic option with said default value.

Reply:
@AHeise , if the user doesn't define the `TOPIC` table option, the value of `defaultTopicId` in the constructor is null and the value of `defaultTopicId` is depended on the `getTargetTopic`. the return value of `getTargetTopic` is depended on the value of `TOPIC` metadata column, and @twalthr mentioned that If `TOPIC` option is not defined any value is supported in the metadata column. For @AHeise 's opinion, if `TOPIC` option is not defined and the value of `TOPIC` metadata column is null, the `defaultTopicId` should be a default value like `unspecified`, right?